src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+                if (0 < offset) {
+                    /* go to next item, updating address */
+                    addr = _get_data_region_by_offset(data_seg, offset);
+                    if (NULL == addr) {
+                        /* report problem and return */
+                        PMIX_ERROR_LOG(PMIX_ERROR);
+                        rc = PMIX_ERROR;
+                        goto done;
+                    }
+                } else {
+                    /* no more data for this rank */
+                    PMIX_OUTPUT_VERBOSE((7, pmix_globals.debug_output,
+                                "%s:%d:%s:  no more data for this rank is found in the shared memory. rank %d key %s not found", __FILE__, __LINE__, __func__, cur_rank, key));
+                    break;
+                }
+            } else if (0 == strncmp((const char *)addr, key, PMIX_MAX_KEYLEN+1)) {
+                PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                            "%s:%d:%s: for rank %s:%d, found target key %s", __FILE__, __LINE__, __func__, nspace, cur_rank, key));
+                /* target key is found, get value */
+                size_t size = *(size_t *)(addr + PMIX_MAX_KEYLEN + 1);
+                addr += PMIX_MAX_KEYLEN + 1 + sizeof(size_t);
+                PMIX_CONSTRUCT(&buffer, pmix_buffer_t);
+                PMIX_LOAD_BUFFER(&buffer, addr, size);
+                int cnt = 1;
+                /* unpack value for this key from the buffer. */
+                PMIX_VALUE_CONSTRUCT(&val);
+                if (PMIX_SUCCESS != (rc = pmix_bfrop.unpack(&buffer, &val, &cnt, PMIX_VALUE))) {
+                    PMIX_ERROR_LOG(rc);
+                } else if (PMIX_SUCCESS != (rc = pmix_bfrop.copy((void**)kvs, &val, PMIX_VALUE))) {
+                    PMIX_ERROR_LOG(rc);
+                }
+                PMIX_VALUE_DESTRUCT(&val);
+                buffer.base_ptr = NULL;
+                buffer.bytes_used = 0;
+                PMIX_DESTRUCT(&buffer);
+                goto done;
+            } else {
+                char ckey[PMIX_MAX_KEYLEN+1] = {0};
+                strncpy(ckey, (const char *)addr, PMIX_MAX_KEYLEN+1);
+                size_t size = *(size_t *)(addr + PMIX_MAX_KEYLEN + 1);
+                PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                            "%s:%d:%s: for rank %s:%d, skip key %s look for key %s", __FILE__, __LINE__, __func__, nspace, cur_rank, ckey, key));
+                /* go to next item, updating address */
+                addr += KVAL_SIZE(size);
+                kval_cnt--;
+            }
+        }
+    }
+
+done:
+    /* unset lock */
+    flock(lockfd, LOCK_UN);
+    return rc;
+}
+
+static void _set_constants_from_env()
+{
+    char *str;
+    int page_size = _pmix_getpagesize();
+    if( NULL != (str = getenv("INITIAL_SEG_SIZE")) ) {
```

Comment:
Constants here must be #define-d and have PMIX_ prefix
We can use this constant to pass `PMIX_SERVER_TMPDIR` value for DSTORE.